compute_sdk/globus_compute_sdk/sdk/diagnostic.py

```diff
@@ -0,0 +1,497 @@
+
+
+def run_all_diags_wrapper(
+    print_to_console: bool, log_bytes: int, ep_uuid: str | None = None
+):
+    """
+    This is the diagnostic wrapper that obtains the list of commands to be run
+    (which depends on whether we are in the SDK or Endpoint, and filters based
+    on current platform OS), then runs the individual diagnostic methods one by
+    one, and collates them into a zip file upon completion.
+
+    For each diagnostic, this method prints what is being run to stdout, then
+    executes the diagnostic itself which may also print to stdout if the flag
+    is specified.  All output is added to a file as tests are run, and the
+    output file is compressed at the conclusion of the method.
+
+    :param print_to_console:  Whether to print the diagnostic output to console
+                                as they are being run.  If False, will only
+                                print the name of each step as they are run
+    :param log_bytes:         # of bytes to limit the individual endpoint log
+                                extractions to
+    :param ep_uuid:           # If specified, register a sample function and
+                                send a task that uses the function UUID to the
+                                endpoint
+    """
+
+    current_date = dt.now().strftime("%Y-%m-%d")
+    log_filename = OUTPUT_FILENAME.format(current_date)
+    zip_filename = f"{log_filename}.gz"
+
+    # Initialize empty file
+    open(log_filename, "w").close()
+
+    diag_cmds, ep_install_dir = get_diagnostic_commands(log_bytes)
+    for cmd in diag_cmds:
+        if cmd is None:
+            # None signifies incompatibility with current OS with no warning needed
+            # Other commands may also be unavailable, but will emit a warning
+            continue
+        elif callable(cmd):
+            display_name = getattr(cmd, "display_name", f"{cmd.__name__}()")
+            display_name = f"python:{display_name}"
+        else:
+            display_name = str(cmd)
+
+        # This displays the command being run to console for every command
+        # diagnostic_display = f"== Diagnostic: {display_name} =="
+        print_highlight(f"== Diagnostic: {display_name} ==")
+
+        append_diagnostic_result(display_name, log_filename, True)
+        with open(CUR_CMD_OUTPUT_FILENAME, "w") as f:
+            with contextlib.redirect_stdout(f):
+                try:
+                    if callable(cmd):
+                        cmd_output = cmd()
+                        # Some methods return a str output, not printing to stdout
+                        if cmd_output is not None:
+                            print(cmd_output)
+                    else:
+                        run_single_command(cmd)
+                except Exception as e:
+                    print(f"Error encountered during ({display_name}): {e}")
+
+        append_diagnostic_result(None, log_filename, console=print_to_console)
```

Comment:
We seem to do more filesystem operations here than are necessary. I suggest we use the original methodology of sending everything to stdout and redirecting to a file, if specified. Alternatively, if we don't want to print to the console by default, we could simply append the diagnostic results to a string, then either write to a file, print to console, or both.

Reply:
> I don't have an option to generate diag txt withOUT

Hmm; thinking about this ~now, I can see a workflow where there's no need or desire to create a file.  Maybe `-p` becomes an either/or flag?  That is, by _default_ we create an archive:

```console
$ globus-compute-diagnostic
Successfully created globus_compute_diagnostic_2024-10-01.txt.gz
```

But if `-p` is utilized, we _strictly_ emit to the console:
```console
$ globus-compute-diagnostic -p
== Diagnostic: uname -a ==
Linux klobble 6.8.0-40-generic #40~22.04.3-Ubuntu SMP PREEMPT_DYNAMIC Tue Jul 30 17:30:19 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

== Diagnostic: python:cat(/etc/os-release) ==
...
 | 1665976858.529456 2022-10-16 23:20:58 INFO MainProce[...] Total number of tasks is 1
 | 
-----------------------------------------------------------------------------

$
```
And _no_ archive file generated.  This leaves open the possibility of someone wanting to pipeline the output to their own tools without cluttering their directory, and does the default action that 98% of folks will use.

Reply:
Addressed in another comment later, after redoing the defaults and options.

Reply:
@khk-globus, your suggestion is mostly in line with the original methodology, except that the default behavior is flipped. I agree with going this route.

My suggestion was to either redirect stdout, as we currently do, or append output to a single string then either print to console or write to a file at the end. I don't see a need for so many FS operations.

Reply:
To summarize, are both of you voting for 'zip file created by default, optional --print flag will print to console instead of creating local file'?

I had already refactored to flip the argument from --print to --zipfile, and now I will go back to the original --print.  Hope this is the final path.

Reply:
I don't have a strong preference for the default behavior, only that we should hold the output in memory and do a single FS operation.

Reply:
Output is in memory and only output at end of diagnostic now.